src/v/cloud_storage_clients/client_pool.cc

```diff
@@ -72,21 +102,62 @@ client_pool::acquire(ss::abort_source& as) {
     client_lease lease(
       client,
       as,
-      ss::make_deleter([pool = weak_from_this(), client, g = std::move(guard)] {
+      ss::make_deleter([pool = weak_from_this(),
+                        client,
+                        g = std::move(guard),
+                        source_sid]() mutable {
           if (pool) {
               pool->release(client);
+              if (source_sid.has_value()) {
+                  // In the background return the client to the connection pool
+                  // of the source shard. The lifetime is guaranteed by the gate
+                  // guard.
+                  ssx::background = pool->container().invoke_on(
+                    source_sid.value(), [g = std::move(g)](client_pool& other) {
+                        other.return_one();
+                    });
+              }
           }
       }));
     _leased.push_back(lease);
+
+    // Update usage stats
+    auto availability = _capacity - std::clamp(_pool.size(), 0UL, _capacity);
+    _availability_stat.update(availability);
```

Comment:
Why do we use history here and not just the current availability? 

Is it possible that history has max availability of N free slots but currently no slots are free, so we end up going over the pool size?

Reply:
I see we check during borrowing that pool is empty so the pool will never be over-allocated. 

But still we might end up with a shard announcing high availability N with lower actual current capacity, based on history; when another shard might have actual higher current capacity N-1, but N gets used? Is this understanding correct?

Reply:
yes, maybe I should get the min(availability, current value)